pkg/core/db.go

```diff
@@ -23,6 +26,23 @@ type NoopDB struct {
 
 // SetUp initializes the database.
 func (NoopDB) SetUp(ctx context.Context, nodes []cluster.Node, node cluster.Node) error {
+	prepareSQL := fixture.Context.TiDBClusterConfig.PrepareSQL
+	if len(prepareSQL) > 0 && node.Component == cluster.TiDB {
+		db, err := sql.Open("mysql", fmt.Sprintf("root@tcp(%s:%d)/test", node.IP, node.Port))
+		if err != nil {
+			return err
+		}
+		if _, err = db.Exec(prepareSQL); err != nil {
+			return err
+		}
+		if err = db.Close(); err != nil {
+			return err
+		}
+		// Wait until global variables take effect
+		time.Sleep(2100 * time.Millisecond)
+		// Only need to run once
+		fixture.Context.TiDBClusterConfig.PrepareSQL = ""
```

Comment:
It seems that it may run for multiple times

```golang
func (c *Controller) syncNodeExec(f func(i int)) {
        var wg sync.WaitGroup
        n := len(c.cfg.Nodes)
        wg.Add(n)
        for i := 0; i < n; i++ {
                go func(i int) {
                        defer wg.Done()
                        f(i)
                }(i)
        }
        wg.Wait()
}

func (c *Controller) setUpDB() {
        log.Infof("begin to set up database")
        c.syncNodeExec(func(i int) {
                log.Infof("begin to set up database on %s", c.cfg.Nodes[i])
                db := core.GetDB(c.cfg.DB)
                err := db.SetUp(c.ctx, c.cfg.Nodes, c.cfg.Nodes[i])
                if err != nil {
                        log.Fatalf("setup db %s at node %s failed %v", c.cfg.DB, c.cfg.Nodes[i], err)
                }
        })
}
```

Reply:
Yes. So we clear this config here to avoid running it multiple times.

Reply:
Emmm, `db.SetUp` runs concurrently. I think it may still run for multiple times even we clear the config here.

Reply:
Ah, I see. It is a race as well.